Marlin/src/module/planner.cpp

```diff
@@ -874,6 +874,10 @@ void Planner::calculate_trapezoid_for_block(block_t * const block, const_float_t
       const float comp = extruder_advance_K[E_INDEX_N(block->extruder)] * block->steps.e / block->step_event_count;
       block->max_adv_steps = cruise_rate * comp;
       block->final_adv_steps = final_rate * comp;
+      #if ENABLED(LA_ZERO_SLOWDOWN)
+        block->la_step_rate = 0;
+        block->max_e_acc = (uint32_t)(float(planner.max_acceleration_steps_per_s2[E_AXIS + E_INDEX_N(extruder)]) * float(block->step_event_count) / float(block->steps[E_AXIS]) * (float(1UL << 24) / (STEPPER_TIMER_RATE)));
```

Comment:
I wonder if this could also be a static variable computed once when the block is dequeued to save more ram. It is a very expensive operation though so it shouldn't go in the isr

Reply:
Advice on where else to put it?

Reply:
I don't think it needs to be calculated in `Planner::calculate_trapezoid_for_block()` which can get called multiple times for any given block. All the variables it depends on are set in stone before or during `Planner::_populate_block()`. I think the logical place for it is in the `#if ENABLED(LIN_ADVANCE)` block at around line 2377.

It can't be static because max acceleration could change between blocks.

Reply:
I think scaling by `float(block->step_event_count) / float(block->steps[E_AXIS])` is incorrect. `block->max_e_acc` is going to be multiplied by an interval measured in stepper timer ticks and the result should be delta speed for the E axis. So it just needs to be steps / s / tick, which you would get without the scaling.

Reply:
I intended to scale it to movement steps (because _"the Bresenham algorithm will convert this step rate into extruder steps"_).
I inferred it from these snippets that are already in the code:


```cpp
#define LIMIT_ACCEL_FLOAT(AXIS,INDX) do{ \
      if (block->steps[AXIS] && max_acceleration_steps_per_s2[AXIS+INDX] < accel) { \
        const float max_possible = float(max_acceleration_steps_per_s2[AXIS+INDX]) * float(block->step_event_count) / float(block->steps[AXIS]); \
        NOMORE(accel, max_possible); \
      } \
    }while(0)
```

and

```cpp
block->acceleration_rate = (uint32_t)(accel * (float(1UL << 24) / (STEPPER_TIMER_RATE)));
```

Which is then used like this in `Stepper::block_phase_isr` 

```cpp
acc_step_rate = STEP_MULTIPLY(acceleration_time, current_block->acceleration_rate) + current_block->initial_rate;
```

----

Am I seeing it wrong? I did it like that so I can multiply by `interval`, which I believe determines the time until the next call to `block_phase_isr`

Reply:
Oh, you're right.